bam_sort.c

```diff
@@ -1591,13 +1683,17 @@ typedef struct {
 
 static inline int heap_add_read(heap1_t *heap, int nfiles, samFile **fp,
                                 int num_in_mem, buf_region *in_mem,
-                                bam1_tag *buf, uint64_t *idx, sam_hdr_t *hout) {
+                                bam1_tag *buf, template_coordinate_keys_t *keys,
+                                uint64_t *idx, sam_hdr_t *hout) {
     int i = heap->i, res;
     if (i < nfiles) { // read from file
         res = sam_read1(fp[i], hout, heap->entry.bam_record);
     } else { // read from memory
         if (in_mem[i - nfiles].from < in_mem[i - nfiles].to) {
-            heap->entry.bam_record = buf[in_mem[i - nfiles].from++].bam_record;
+            size_t from = in_mem[i - nfiles].from;
+            heap->entry.bam_record = buf[from].bam_record;
+            if (g_sam_order == TemplateCoordinate) heap->entry.u.key = template_coordinate_keys_get(keys, from);
+            in_mem[i - nfiles].from++;
```

Comment:
I'm also getting a segmentation fault from the heap sort stage.  I think the problem is that while you've still got the `template_coordinate_keys_t` data for everything in-memory, unfortunately you've lost it for reads that were spilled to disk.  It would therefore need to be recreated for the reads that are read back in again.  You wouldn't need that many - it should be one for every temporary file if the heap is working correctly.